Userland/Utilities/dns.cpp

```diff
@@ -0,0 +1,132 @@
+                    ByteString rr_name = rr;
+                    auto type = DNS::Messages::resource_type_from_string(rr_name.to_uppercase());
+                    if (!type.has_value())
+                        return Error::from_string_literal("Invalid resource type");
+                    current_request.types.append(type.value());
+                }
+                current_request.name = parts[1];
+            }
+            requests.append(move(current_request));
+            current_request = {};
+            return true;
+        },
+    });
+
+    args_parser.parse(arguments);
+
+    if (server_address.is_empty()) {
+        outln("You must specify a server address to query");
+        return 1;
+    }
+
+    if (current_request.name.is_empty() && !current_request.types.is_empty()) {
+        outln("You must specify at least one DNS record to query");
+        return 1;
+    }
+
+    Core::EventLoop loop;
+
+    DNS::Resolver resolver {
+        [&] -> ErrorOr<DNS::Resolver::SocketResult> {
+            Core::SocketAddress addr;
+            if (auto v4 = IPv4Address::from_string(server_address); v4.has_value()) {
+                addr = { v4.value(), static_cast<u16>(use_tls ? 853 : 53) };
+            } else if (auto v6 = IPv6Address::from_string(server_address); v6.has_value()) {
+                addr = { v6.value(), static_cast<u16>(use_tls ? 853 : 53) };
+            } else {
+                return MUST(resolver.lookup(server_address)->await())->cached_addresses().first().visit([&](auto& address) -> DNS::Resolver::SocketResult {
+                    if (use_tls) {
+                        auto tls = MUST(TLS::TLSv12::connect({ address, 853 }, server_address));
+                        return { move(tls), DNS::Resolver::ConnectionMode::TCP };
+                    }
+                    return { MUST(Core::BufferedSocket<Core::UDPSocket>::create(MUST(Core::UDPSocket::connect({ address, 53 })))), DNS::Resolver::ConnectionMode::UDP };
+                });
+            }
+
+            if (use_tls)
+                return DNS::Resolver::SocketResult { MUST(TLS::TLSv12::connect(addr, server_address)), DNS::Resolver::ConnectionMode::TCP };
+            return DNS::Resolver::SocketResult { MUST(Core::BufferedSocket<Core::UDPSocket>::create(MUST(Core::UDPSocket::connect(addr)))), DNS::Resolver::ConnectionMode::UDP };
+        }
+    };
+
+    DefaultRootCACertificates::set_default_certificate_paths(Array<ByteString, 1> { cert_path.is_empty() ? "/etc/ssl/cert.pem"sv : cert_path });
+
+    MUST(resolver.when_socket_ready()->await());
+
+    size_t pending_requests = requests.size();
+    for (auto& request : requests) {
+        resolver.lookup(request.name, DNS::Messages::Class::IN, request.types)
+            ->when_resolved([&](auto& result) {
+                outln("Resolved {}:", request.name);
+                HashTable<DNS::Messages::ResourceType> types;
+                auto& recs = result->records();
```

Comment:
CI is mad about the auto& here, because `result->records()` returns a copy of the vector of records.

Reply:
Ah, thanks for catching that
I'm surprised it worked on macOS 🤔

Reply:
Oh I think I just tested building ladybird, not the utility